src/plugins/dump/dump.cpp

```diff
@@ -78,125 +111,347 @@ int serialise (std::ostream & os, ckdb::Key *, ckdb::KeySet * ks)
+	ss >> command;
+
+	if (command == "kdbOpen")
+	{
+		std::string version;
+		ss >> version;
+		if (version != "1")
+		{
+			ELEKTRA_SET_INSTALLATION_ERRORF (parentKey, "Wrong version detected in dumpfile: %s", version.c_str ());
+			return -1;
+		}
+	}
+	else if (command == "ksNew")
+	{
+		ss >> nrKeys;
+
+		ksClear (ks);
+	}
+	else if (command == "keyNew")
+	{
+		cur = ckdb::keyNew (nullptr);
+
+		ss >> namesize;
+		ss >> valuesize;
+
+		if (namesize > namebuffer.size ()) namebuffer.resize (namesize + 1);
+		is.read (&namebuffer[0], namesize);
+		namebuffer[namesize] = 0;
+		ckdb::keySetName (cur, &namebuffer[0]);
+
+		if (valuesize > valuebuffer.size ()) valuebuffer.resize (valuesize + 1);
+		is.read (&valuebuffer[0], valuesize);
+		valuebuffer[valuesize] = 0;
+
+		ckdb::keySetRaw (cur, &valuebuffer[0], valuesize);
+		std::getline (is, line);
+	}
+	else if (command == "keyMeta")
+	{
+		ss >> namesize;
+		ss >> valuesize;
+
+		if (namesize > namebuffer.size ()) namebuffer.resize (namesize + 1);
+		is.read (&namebuffer[0], namesize);
+		namebuffer[namesize] = 0;
+
+		if (valuesize > valuebuffer.size ()) valuebuffer.resize (valuesize + 1);
+		is.read (&valuebuffer[0], valuesize);
+		valuebuffer[valuesize] = 0;
+
+		keySetMeta (cur, &namebuffer[0], &valuebuffer[0]);
+		std::getline (is, line);
+	}
+	else if (command == "keyCopyMeta")
+	{
+		ss >> namesize;
+		ss >> valuesize;
+
+		if (namesize > namebuffer.size ()) namebuffer.resize (namesize + 1);
+		is.read (&namebuffer[0], namesize);
+		namebuffer[namesize] = 0;
+
+		if (valuesize > valuebuffer.size ()) valuebuffer.resize (valuesize + 1);
+		is.read (&valuebuffer[0], valuesize);
+		valuebuffer[valuesize] = 0;
+
+		ckdb::Key * search = ckdb::ksLookupByName (ks, &namebuffer[0], 0);
+		ckdb::keyCopyMeta (cur, search, &valuebuffer[0]);
+		std::getline (is, line);
+	}
+	else if (command == "keyEnd")
+	{
+		ckdb::ksAppendKey (ks, cur);
+		cur = nullptr;
+	}
+	else if (command == "ksEnd")
+	{
+		return 1;
+	}
+	else
+	{
+		ELEKTRA_SET_VALIDATION_SYNTACTIC_ERRORF (
+			parentKey,
+			"Unknown command detected in dumpfile: %s.\nMaybe the file is not in dump configuration file format? "
+			"Try to remount with another plugin (eg. ini, ni, etc.)",
+			command.c_str ());
+		return -1;
+	}
+
+	*curPtr = cur;
+
+	return 0;
+}
+
+int unserialiseVersion1 (std::istream & is, ckdb::Key * parentKey, ckdb::KeySet * ks, const std::string & firstLine)
+{
+	ckdb::Key * cur = nullptr;
+	std::string line = firstLine;
+
+	do
+	{
+		int ret = decodeLine (is, parentKey, ks, line, &cur);
+
+		if (ret == -1)
+		{
+			return ELEKTRA_PLUGIN_STATUS_ERROR;
+		}
+
+		if (ret == 1)
+		{
+			break;
+		}
+	} while (std::getline (is, line));
+
+	return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+}
+
+int unserialiseVersion2 (std::istream & is, ckdb::Key * parentKey, ckdb::KeySet * ks, bool useFullNames)
+{
+	ckdb::Key * cur = nullptr;
+	std::string line;
+
+	std::string rootName (keyName (parentKey));
+	rootName += "/";
+
+	if (useFullNames)
```

Comment:
I do not like this option, why not always use relative names? (And iterate over all namespaces where needed, like with pluginprocess.)

Reply:
The problem is that `pluginprocess` wants to send a `KeySet` with `Key`s that are **not** below the `parentKey`. This is very weird and I am totally in favour of changing it, but I didn't want to touch the convoluted `pluginprocess` code more than necessary.

I intentionally left this option undocumented, so that it could be removed easily once `pluginprocess` is implemented properly. Ideally `pluginprocess` should switch to the much faster `quickdump` anyway, but that requires changes to `quickdump` to make streaming possible.

> And iterate over all namespaces where needed, like with pluginprocess.

Not sure what you mean...

Reply:
This was an implementation hint how the pluginprocess communication needs to be rewritten. So instead of sending a single KeySet, we would need a protocol outside of KeySets which tells from which namespace the next KeySet will be. Such a protocol would have been needed for all other plugins that are relative anyway. And now all plugins are relative #51. (Actually this PR really closes #51.)

But yes, it can (probably should) be done in a different PR. Can you open an issue about it?

Reply:
I'll open a new issue. `pluginprocess` probably should be rethought entirely. It probably shouldn't rely on a storage plugin to write the `KeySet` it wants to send. It should just write/read the keys directly. Maybe we could even use `mmapstorage` with a shared memory approach and avoid the pipes entirely. The pipes seem to cause problems on macOS anyway (there is another workaround for that in `dump`).

Reply:
I like the idea with mmapstorage! @mpranj what do you think?

Reply:
Discussion moved to #3509